bloom-works/guides-template
# Add ESLinter for JS Linting

This changes is related to the [Setup Guides](https://github.com/bloom-works/guides-template/issues/5) task.  It adds linting functionality for JavaScript files.  The linter can be run at any point during development.  It is also added to the workflow process (from PR click checks tab in top-middle of screen, click Pull Request on left side of screen, click pr_checks on left-side or middle of screen, click Run Linter in middle of screen).  Prior to this change the run linter option in PR Check was a print statement as a placeholder.  The image below shows a screenshot of the workflow with ESLint running and passing during the PR check of this current PR.
<img width="930" alt="Screenshot 2023-07-04 at 8 53 06 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/a517be16-b0db-4d2e-b8b6-b92b089d684a">


To make it more consumable for review, I will split up all the linter work into PRs associated to each linter and language.  This review will include all changes for ESLint and JavaScript code.

### What was changed

- Added ESLint capabilities (packages, configuration)
- Updated sh file used in github workflows for pr check 
- Add steps to run before PR to developer section of README 
- Fixed changes in JS files reported by linter

### How to check change

1. Checkout branch and build project
2. Verify all changes reported by linter did not disrupt the project by running `npm run serve`
3. Verify there are no existing linter errors by using the path on PR above to see if pr checks passed for linter
4. Verify there are no existing linter errors by manually running the linter in the code.  Navigate to the base of the repo and run `sh ./ci/linter.sh`.  Below image shows an example of what the command line will look like when there are no errors.
<img width="311" alt="Screenshot 2023-07-04 at 8 27 28 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/f38ba5e3-4409-46f6-aaff-e0dad41e2631">

5. Verify linter is working and/or check any style fixes I made in other files.  For this you will use the same command as above and choose any of the files that have changed and reverse any changes made.  I have give two examples below, but feel free to check any of the changes using the same process.

Example 1: Here I have changed the single quotes back to double quotes on line 2 of the markdown.js file and then run the linter.  Note the linter now reports an error for _quotes_ with the message that _Strings must use singlequote_.
<img width="576" alt="Screenshot 2023-07-04 at 8 28 41 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/6dbb6fe6-4f62-4248-8eca-acd5739cd616">
 
Example2: Here I have navigated to a different directory from the first example (_includes/components/) and removed the added space on line 4 of the PageHeader.js file.  Note the linter now reports and error for _space-before-function-paren_ with the message _Missing space before function parentheses_.
<img width="583" alt="Screenshot 2023-07-04 at 8 29 48 AM" src="https://github.com/bloom-works/guides-template/assets/137226010/2b696249-8282-4615-9833-27b152d5cb90">


### Notes

I have turned two rules off within the .eslintrc.js file. 

- semi: By default this will error for any added semicolons at the end of lines.  The semicolon is not required in JS but can be useful to avoid syntax errors when the compiler cannot auto-add.  We have a mixed usage, which _could_ be intentional.  If we ever decide to always use or never use, I have added a note to the configuration on what to change it to.
- no-unused-vars: I believe this is due to not all changes being added to the project yet and we can remove this ignored rule completly once all changes have been merged to main.
